### PR TITLE
Improve server loop

### PR DESCRIPTION
**Pull request by smondet**

### CODE REVIEW

File: src/lib/command_line.ml [hunk 1]
```diff
@@ -66,11 +66,17 @@ module Server = struct
         ]
       | `Loop_ends errors ->
         "loop",
-        json_event "loop-begins" [
+        json_event "loop-ends" [
           "errors", `List (List.map errors ~f:(fun x ->
               `String (Error.to_string x)));
           "jobs", current_jobs ();
         ]
+      | `Incomming_job job ->
```
Comment (ihodes) [inline src/lib/command_line.ml]: incoming*

File: src/lib/command_line.ml [hunk 2]
```diff
@@ -80,10 +81,25 @@ module Server = struct
...
-  let save_job_list t =
-    Storage.Json.save_jsonable t.storage
-      ~path:["server"; "jobs.json"]
-      (`List (List.map t.jobs ~f:(fun j -> `String (Job.id j))))
+  let change_job_list t action =
+    let mutex =
+      match t.job_list_mutex with
+      | Some s -> s
+      | None ->
+        let m = Lwt_mutex.create () in
```
Comment (ihodes) [inline src/lib/command_line.ml]: Shouldn't this be initialized at server init time? This looks like a potential race condition (two calls to this fn could result in the first call to hold an invalid, overwritten mutex)? I don't know how this is called though, so maybe it's never called in parallel (but then why a mutex?) Why do we need a mutex?

File: src/lib/log.mli [hunk 3]
```diff
@@ -10,3 +10,12 @@ val log :
...
   (unit,
    [> `Storage of [> Storage.Error.common ] ]) Deferred_result.t
+
+
+val debug_sections : string list list ref
+(** Catch sections and display them on ["stdout"], cf. the [?section] argument of {!log}.
+
+    At the start of the program, the [Log] module tries to parse
+    the ["COCLOBAS_DEBUG_SECTIONS"] environment variable to fill this
+    reference (coma-separated list of paths, e.g. ["/server/loop,cluster//commands"]).
```
Comment (ihodes) [inline src/lib/log.mli]: comma*

File: src/lib/log.ml [hunk 4]
```diff
@@ -11,6 +11,15 @@ type t = [
 let silent : t = `Silent
 let stored storage : t = `Stored {storage}
 
+let debug_sections : string list list ref = ref []
+let () =
+  try debug_sections :=
```
Comment (ihodes) [inline src/lib/log.ml]: nice

**Comment by ihodes** [comment]
LGTM overall, couple questions. 

Might be time to move `Server` to its own file? :)

File: src/lib/command_line.ml (hunk 2 above)
Comment (smondet) [inline src/lib/command_line.ml]: `Lwt` is cooperative so there cannot be a preemption between the beginning of the function and the call to `Lwt_mutex.with_lock`.

The mutex is there to make sure updates `t.jobs <- ...` happen together with the `Storage.save_jsonable`.

I had to convince myself again ☺, so maybe I'll try to come up with something more readable/obvious.

Comment (ihodes) [inline src/lib/command_line.ml]: > The mutex is there to make sure updates t.jobs <- ... happen together with the Storage.save_jsonable.

If there weren't a mutex, how would this become unsequenced? Isn't the server loop sequential? (I believe you, just trying to figure out where my mental map is going wrong).

Comment (ihodes) [inline src/lib/command_line.ml]: Ah I got it; my confusion was in 1. forgetting that Lwt uses cooperative threading and 2. not thinking to check that `Storage.save_jsonable` used a `Lwt.bind`, which is where the problem could've occurred. Thank you for the clarification :)

**Comment by ihodes** [comment]
LGTM!

**Pull request merged**